### PR TITLE
Gallery block: remove the v1 code and add deprecations to auto transform the v1 galleries

### LINKED ISSUE #32075: Gallery block: automatically migrate v1 blocks to v2 on edit
At some point in the future we will want to automatically migrate the v1 block content to v2 format on edit and remove the v1 edit and save code.

Some work was done previously on setting up the deprecations for this, but the code was removed when the decision was made to continue supporting the v1 format. This removed code can be found at https://gist.github.com/glendaviesnz/b295c625e4dfe59c678ae396f5c361d5 or at the original commit https://github.com/WordPress/gutenberg/blob/0ff156a277ae6cb7589ca016addb80f8a6a628de/packages/block-library/src/gallery/deprecated.js

Also, the changes needed to the underlying fixtures, etc. can be found at https://github.com/WordPress/gutenberg/pull/28481. This will likely be outdated by the time it is reimplemented, but may provide information on what is required in order to update fixtures to new version of the block.

### PR DESCRIPTION
**Pull request by glendaviesnz**
## Description

Removes the gallery v1 code and adds deprecations to auto upgrade v1 galleries on edit

Fixes: #32075

## Testing

- Before checking out the PR, on an existing local dev env disable the gallery experiment and add a number of v1 galleries of different sizes, column numbers, linkTo and image size options
- Check out this PR and enable the gallery experiment
- Load the previously created galleries in the editor and make sure they are automatically migrated to the new format correctly
- Make sure that `npm run text-unit` and `npm run fixtures:regenerate` complete successfully
- Test manual transformations to and from Jetpack gallery and slideshow blocks, and from gallery shortcodes (easy way to get valid shortcodes is to add a gallery in the classic block)

### CODE REVIEW

**Comment by ramonjd** [comment]
I've noticed that attachment links aren't migrated for galleries where ~~I'm linking to the attachment and setting an image size~~. I'm linking to the attachment for a single image first, then the whole gallery.

_Edit_: I been repeating the experiment from scratch and can't reproduce again. Sorry about the red herring.

**Comment by ramonjd** [comment]
I've been testing the following:

- Core Gallery with varying columns, sizes, linkTo and image size values
- Jetpack Tiled Gallery
- Jetpack Slideshow

After switching branches and transformation, I don't notice anything strange and the galleries appear as they should.

The Core Galleries's blocks are transforming correctly. 👍 

I did manage to see the [issue reported above](https://github.com/WordPress/gutenberg/pull/35609#issuecomment-943871578) in a few cases, mostly when I had multiple core galleries on the same page. The transformed gallery's attributes look like `<!-- wp:gallery {"columns":3,"linkTo":"attachment"} -->` but there are no attachment links on the frontend. Sorry I haven't managed to reliably reproduce, hopefully someone might provide a confidence check on this. 🤞 

I'm not sure what to expect to see for the Jetpack Galleries. I didn't notice any transformations, however the galleries themselves were unaffected.

**Comment by glendaviesnz** [comment]
> I'm not sure what to expect to see for the Jetpack Galleries. I didn't notice any transformations, however the galleries themselves were unaffected.

The Jetpack blocks should have an option to manually 'transform' to Gallery, and if selected should transform to the new gallery format.

**Comment by ramonjd** [comment]
> The Jetpack blocks should have an option to manually 'transform' to Gallery, and if selected should transform to the new gallery format.

Thanks. 🤦 Should have read the instructions better. Just retested Tile Gallery manual transforms. They transform to the new core gallery as expected ✅

**Comment by glendaviesnz** [comment]
> in a few cases, mostly when I had multiple core galleries on the same page. The transformed gallery's attributes look like `<!-- wp:gallery {"columns":3,"linkTo":"attachment"} -->` but there are no attachment links on the frontend.

🤞 this should be fixed by this commit https://github.com/WordPress/gutenberg/pull/35609/commits/28bf0f604e070f2ee38e2493907fe46d4d643eff

File: packages/block-library/src/gallery/deprecated.js
Comment (andrewserong) [inline packages/block-library/src/gallery/deprecated.js]: Tiniest of tiny nits, there's an additional whitespace character introduced here so the alignment is slightly off.

File: packages/block-library/src/gallery/deprecated.js
Comment (andrewserong) [inline packages/block-library/src/gallery/deprecated.js]: Just a tiny typo:

```suggestion
 * Original function to determine new href and linkDestination values for an image block from the
```

File: packages/block-library/src/gallery/deprecated.js
Comment (andrewserong) [inline packages/block-library/src/gallery/deprecated.js]: Tiny nit: there's an additional white space character on the return type label
```suggestion
 * @return {Object}         Image block.
```

File: packages/block-library/src/gallery/deprecated.js
Comment (andrewserong) [inline packages/block-library/src/gallery/deprecated.js]: Is this missing returning `innerBlocks`? Each of the other deprecations' `migrate` functions are returning innerBlocks, too. I assume this means we might need to add another set of fixtures for this deprecation, too?

**Review by andrewserong — COMMENTED** [review]
Great work @glendaviesnz! This is testing nicely, all looked good to me except for one of the migrate functions.

✅ After applying the PR, non migrated gallery blocks (including shortcode base gallery) render as before on the front end of the site
✅ JP Tiled gallery -> Gallery block transform works correctly
✅ JP Slideshow -> Gallery block transform works correctly
✅ Gallery shortcode -> Gallery block transform works correctly
✅ v1 Gallery -> Image based gallery block transform works correctly
✅ Link to "Attachment Page" setting transferred correctly
✅ Link to "Media file" setting transferred correctly
✅ Image size setting transferred correctly
✅ Visually, the code change for the transforms looks consistent (removes the transforms for the v1 gallery, and the transforms to the image-based gallery look correct) 
✅ As far as I can tell from skimming the fixtures, the changes look good
✅ Edit Wrapper removal looks good, as does the change to the save method
❓ Most of the changes in `deprecated.js` look good to me. I added a couple of very minor whitespace comments, but the main issue I noticed is that the `v2` deprecation doesn't return `innerBlocks` in the migrate function, like the other deprecations do. Because all the tests are currently passing, I was wondering if we need to add another deprecation to catch this one, too?

I'm a little fuzzy still about how the `migrate` function works, so apologies if it's a red herring!

Other than that, this PR looks really good. I imagine we'll need to hold off on merging this until the update to `nestable_tags` is merged in core (https://github.com/WordPress/wordpress-develop/pull/1678)?

File: packages/block-library/src/gallery/deprecated.js
Comment (glendaviesnz) [reply packages/block-library/src/gallery/deprecated.js]: Turned out that the v2 fixture was falling through to v3 - have combined into one as were identical apart from `ids` attribute which is stripped out in v7 anyway.

**Comment by ramonjd** [comment]
> 🤞 this should be fixed by this commit 28bf0f6

Thanks, @glendaviesnz I've tried to reproduce the original error, but it can't anymore. 👍

**Review by andrewserong — APPROVED** [review]
Thanks for merging the v2 and v3 deprecations @glendaviesnz (and nice work managing the complexity of that file)! That's testing well for me, and looking at the diff between the two versions, it looks like the merged one handles both versions of the `save` function correctly because of how the attributes are sourced from the `img` blocks in those deprecations. I've thrown a bunch of different potential gallery markup (manually creating variations from the existing test fixtures), and each permutation gets migrated correctly 👍

This PR looks good to me, it's testing well, and I haven't found any other issues with the code change, so I'm giving it the ✅ . I imagine we won't be able to merge it until https://github.com/WordPress/wordpress-develop/pull/1678 lands in core, since this PR removes the `use_BalanceTags` check that will be introduced in https://github.com/WordPress/gutenberg/pull/34979?

**Review by mkevins — APPROVED** [review]
I have tested these changes on mobile using various combinations of original-format gallery content (with captions, and various settings) and everything auto-converts to the new format as expected. Nice work Glen!

**Review by aristath — APPROVED** [review]

**Comment by glendaviesnz** [comment]
The plan is to merge this into 12.0 and cherry pick into 5.9 so don't merge until after 11.9 release.

**Comment by noisysocks** [comment]
> The plan is to merge this into 12.0 and cherry pick into 5.9 so don't merge until after 11.9 release.

No problem. Leave it with me, I can handle that.

**Comment by glendaviesnz** [comment]
We need to keep the block on sites using `use_balanceTags` in place for sites in WP < 5.9 so closing this in favour of https://github.com/WordPress/gutenberg/pull/36191

**Pull request closed without merging**
